server/src/main/java/io/crate/execution/ddl/tables/AddColumnRequest.java

```diff
@@ -0,0 +1,286 @@
+            geoProperties.put("precision", geoRef.precision());
+            geoProperties.put("tree_levels", geoRef.treeLevels());
+            geoProperties.put("distance_error_pct", geoRef.distanceErrorPct());
+        }
+        return new StreamableColumnInfo(
+            reference.column().name(),
+            reference.column().fqn(),
+            reference.valueType(),
+            reference.position(),
+            false, // A primary key can be specified only on ADD COLUMN
+            true, // not-null can be specified only on ADD COLUMN
+            true, // Use default, can be specified only on ADD COLUMN
+            reference.indexType(),
+            reference.valueType().id() == ArrayType.ID,
+            null, // INDEX USING FULLTEXT can be specified only on ADD COLUMN
+            null, // generated expression can be specified only on ADD COLUMN
+            reference.columnPolicy(),
+            geoTree,
+            geoProperties,
+            List.of(), // copyToTargets is used when INDEX is specified, only for ADD COLUMN
+            List.of() // there is no tree-like Ref hierarchy in target.valueType().valueForInsert(row.get(i)), it's a single map
+        );
+    }
+
+    /**
+     * For ADD COLUMN.
+     */
+    public AddColumnRequest(RelationName relationName,
+                            AnalyzedColumnDefinition colToAdd,
+                            Map<String, String> checkConstraints) {
+        this.relationName = relationName;
+        this.columns.add(new StreamableColumnInfo(colToAdd));
+        this.checkConstraints.addAll(checkConstraints.entrySet()
+            .stream()
+            .map(nameExprEntry -> new StreamableCheckConstraint(nameExprEntry.getKey(), nameExprEntry.getValue()))
+            .collect(Collectors.toList()));
+    }
+
+    public RelationName relationName() {
+        return this.relationName;
+    }
+
+    public List<StreamableCheckConstraint> checkConstraints() {
+        return this.checkConstraints;
+    }
+
+    public List<StreamableColumnInfo> columns() {
+        return this.columns;
+    }
+
+    @Override
+    public void writeTo(StreamOutput out) throws IOException {
+        super.writeTo(out);
+        relationName.writeTo(out);
+
+        out.writeVInt(checkConstraints.size());
+        for (int i = 0; i < checkConstraints.size(); i++) {
+            checkConstraints.get(i).writeTo(out);
+        }
+
+        out.writeVInt(columns.size());
+        for (int i = 0; i < columns.size(); i++) {
+            columns.get(i).writeTo(out);
+        }
+    }
+
+    /**
+     * @param name is a non-fqn name of a column.
+     * @param fqn is not streamed but constructed in readFrom() based on name and hierarchy.
+     * We create StreamableColumnInfo from the root, name == fqn on the first call and then update prefix on traversing children.
+     * FQN is available when created from AnalyzedColumnDefinition.
+     */
+    public record StreamableColumnInfo(@Nonnull String name,
```

Comment:
Somehow `Reference` already is streamable and used to represent a typed column.

I currently don't have a better solution, but I think this is a sign that our data structures are not ideal.

- `ColumnDefinition`
- `AnalyzedColumnDefinition`
- `Reference`
- Now `StreamableColumnInfo`

What's also a bit strange is that `Reference` is currently not hierarchical (which sometimes is a useful), yet it's `DataType` with `ObjectType` can be.

I kinda wish we'd find a way to streamline this a bit and have fewer structures to represent columns. E.g. ideally go from `ColumnDefinition` to `XYColumn` and have `XYColumn` implement `Reference` or something like that.

Reply:
To clarify: This is probably not a blocker, but I think it might be worth exploring the design space a bit more before we commit to this.

Reply:
>What's also a bit strange is that Reference is currently not hierarchical
From what I have seen in runtime, whenever we need hierarchy we don't use `Reference`.
- On table creation/col addition it's `AnalyzedColumnDefinition` which has children and is converted to a `Map`.
- In dynamic mapping updates, `valueForInsert`  in `generateSourceAndCheckConstraints` is a map (source to be parsed?) but target reference itself is a flat and this source map used in `SchemaUpdateRequest`.

> XYColumn implement Reference

maybe other way around would fit better as `Reference` has redundant things which are not needed on `ADD COLUMN `and/or dynamic mapping updates, say `defaultExpression` and we don't want a column (at least in add col context) have it.

just thinking loudly:

Our request must convey tree structure. Actually, `Reference` can be used to represent a hierarchy since it has `ColumnIdent` which has path. 

To resolve issue with redundancy of `Reference` for ADD COLUMN we can group columns, relevant for add column into subclass and make it streamable, so that this sub-reference fields group would be pluggable into the request. And There is already `StreamableColumnInfo` (better name it `AddColumnInfo` or smth like that?) and we can put it inside `Reference`. We would also remove `children` field from `StreamableColumnInfo` and put there `ColumnIdent` to represent tree via `column.path`.


However, it's not optimal for streaming as full path of deepest tree is superset of all parent's paths, for example, a->b->c tree would have `a` listed twice as a part of `b.path` and `c.path`

Reply:
> has redundant things which are not needed on ADD COLUMN 

Which ones? `RowGranularity` could be a constant. That leaves `ReferenceIdent`. Maybe worth double-checking if this is still needed, or if we could get rid of that.

Reply:
`defaultExpression` is not supported on add column. Should we just move it out to column then?